scispacy/candidate_generation.py

```diff
@@ -0,0 +1,307 @@
+    # We need to keep around a map from text to possible canonical ids that they map to.
+    text_to_concept_id: Dict[str, Set[str]] = defaultdict(set)
+
+    for concept in umls:
+        for alias in set(concept["aliases"]).union({concept["canonical_name"]}):
+            text_to_concept_id[alias].add(concept["concept_id"])
+
+    # nmslib hyperparameters (very important)
+    # guide: https://github.com/nmslib/nmslib/blob/master/python_bindings/parameters.md
+    # Default values resulted in very low recall.
+
+    # set to the maximum recommended value. Improves recall at the expense of longer indexing time.
+    # TODO: This variable name is so hot because I don't actually know what this parameter does.
+    m_parameter = 100
+    # `C` for Construction. Set to the maximum recommended value
+    # Improves recall at the expense of longer indexing time
+    construction = 2000
+    num_threads = 60  # set based on the machine
+    index_params = {'M': m_parameter, 'indexThreadQty': num_threads, 'efConstruction': construction, 'post' : 0}
+
+    print(f'No tfidf vectorizer on {tfidf_vectorizer_path} or ann index on {ann_index_path}')
+    umls_concept_aliases = list(text_to_concept_id.keys())
+
+    umls_concept_aliases = numpy.array(umls_concept_aliases)
+
+    print(f'Fitting tfidf vectorizer on {len(umls_concept_aliases)} aliases')
+    tfidf_vectorizer = TfidfVectorizer(analyzer='char_wb', ngram_range=(3, 3), min_df=10, dtype=numpy.float32)
+    start_time = datetime.datetime.now()
+    uml_concept_alias_tfidfs = tfidf_vectorizer.fit_transform(umls_concept_aliases)
+    print(f'Saving tfidf vectorizer to {tfidf_vectorizer_path}')
+    joblib.dump(tfidf_vectorizer, tfidf_vectorizer_path)
+    end_time = datetime.datetime.now()
+    total_time = (end_time - start_time)
+    print(f'Fitting and saving vectorizer took {total_time.total_seconds()} seconds')
+
+    print(f'Finding empty (all zeros) tfidf vectors')
+    empty_tfidfs_boolean_flags = numpy.array(uml_concept_alias_tfidfs.sum(axis=1) != 0).reshape(-1,)
+    deleted_aliases = umls_concept_aliases[empty_tfidfs_boolean_flags == False] # pylint: disable=singleton-comparison
+    number_of_non_empty_tfidfs = len(deleted_aliases)
+    total_number_of_tfidfs = uml_concept_alias_tfidfs.shape[0]
+
+    print(f'Deleting {number_of_non_empty_tfidfs}/{total_number_of_tfidfs} aliases because their tfidf is empty')
+    # remove empty tfidf vectors, otherwise nmslib will crash
+    umls_concept_aliases = umls_concept_aliases[empty_tfidfs_boolean_flags].tolist()
+    uml_concept_alias_tfidfs = uml_concept_alias_tfidfs[empty_tfidfs_boolean_flags]
+    print(deleted_aliases)
+
+    print(f'Saving list of concept ids and tfidfs vectors to {uml_concept_aliases_path} and {tfidf_vectors_path}')
+    json.dump(umls_concept_aliases, open(uml_concept_aliases_path, "w"))
+    scipy.sparse.save_npz(tfidf_vectors_path, uml_concept_alias_tfidfs.astype(numpy.float16))
```

Comment:
or was it only here that we can change it to float16

Reply:
I'll add a comment here, basically we can only serialize in float16